[WIP] Template Loader

vicb opened the pull request:
Work in progress for the Template Loader.

What is implemented:
- `TemplateLoader.load()` returns a Promise & `Compiler.compile()` is async,
- The shadow DOM strategy is now global to the application (vs component specific before)

TODO:
- [x] extract the styles from the templates when the emulated strategy is used (insert them in the `<head>`),
- [x] html import based template loader

vsavkin committed:
feat(change_detection): update change detection benchmark

kevmoo committed:
more types in fascades

vicb committed:
feat(compiler): Async compile

vicb committed:
refactor(ShadowDom): Make the strategy global to the application

vsavkin commented:
I am probably missing something, but it looks like you are making the strategy global. This means that you won't be able to use emulated and shadow-dom components in the same application. I think we need this capability.

vicb commented:
Victor, that's right.
Having a global strategy is what we have discussed with Misko and Tobias yesterday.
Why do you think we need to mix both strategies ?

Le 30 janvier 2015 20:47:45 GMT+01:00, Victor Savkin <notifications@github.com> a écrit :
>I am probably missing something, but it looks like you are making the
>strategy global. This means that you won't be able to use emulated and
>shadow-dom components in the same application. I think we need this
>capability.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/angular/angular/pull/499#issuecomment-72259261

vsavkin committed:
feat(change_detection): add mode to ChangeDetector

vicb committed:
feat(ShadowDom): Shim css for emulated shadow dom

vicb committed:
feat(TemplateLoader): support xhr loading

vsavkin committed:
feat(change_detection): do not register a change from switching from null to null

rkirov commented on modules/core/src/application.js:
```diff
@@ -52,12 +56,21 @@ function _injectorBindings(appComponentType) {
 
-      bind(appViewToken).toAsyncFactory((changeDetection, compiler, injector, appElement,
-            appComponentAnnotatedType) => {
+      bind(appViewToken).toAsyncFactory((changeDetection,
```
revert back to argument-packing indentation style. We have to use consistent style in the codebase and argument-packing vs new-line style is preferred by most teammembers.

I don't want to start a big debate on the merits of each, but we have to be consistent.

rkirov commented on modules/core/src/compiler/compiler.js:
```diff
@@ -53,59 +50,104 @@ export class Compiler {
+      // The protoView is resolved after all the components in the template
+      // have been compiled.
+      return PromiseWrapper
+        .all(componentPromises)
```
How does `all` handle error handling. Would this promise reject if one of its dependents fail (not all)?

vicb replied:
It fails if any of the dependents fails

rkirov commented on modules/core/src/compiler/compiler.js:
```diff
@@ -53,59 +50,104 @@ export class Compiler {
+
+    return pvPromise;
+  }
+
```
What happens when a `tmpLoader` fails? We might need to attach `pvPromise.catch()` callback to have a more descriptive erorr.

vicb replied:
I plan to add more descriptive errors

rkirov commented on modules/core/src/compiler/template_loader.js:
```diff
@@ -1,11 +1,52 @@
 export class TemplateLoader {
-  load(url:string):Promise<Document> {
-    return null;
+  // todo(vicb): was Promise<Document>
```
change this `todo` to something more descriptive. It doesn't tell me what you have `to do` :)

vicb replied:
I'll update this.
We can only use a `Document` with html import and the preferred way for now will be to use XHR

rkirov commented on modules/core/src/compiler/template_loader.js:
```diff
@@ -1,11 +1,52 @@
+    if (isPresent(tplConfig.url)) {
+      var url = tplConfig.url;
+      var completer = PromiseWrapper.completer();
+      var link = DOM.createElement('link');
```
What would be the alternative to html5 imports? According to http://caniuse.com/#search=imports we going to miss evergreen Firefox and iOS Safari. Add a comment to remind us what to do.

vicb replied:
As discussed with Tobias and Misko, XHR will be preferred for now

rkirov commented on modules/core/src/compiler/template_loader.js:
```diff
@@ -1,11 +1,52 @@
+      var link = DOM.createElement('link');
+      link.rel = 'import';
+      link.href = url;
+      DOM.on(link, 'load', function() {
```
use fat errors `=> ()` ;)

vicb replied:
Seems like (from a PR comment) `function` is faster when you don't need to bind this (probably applicable to ES5 only)

rkirov commented on modules/core/test/compiler/compiler_spec.js:
```diff
@@ -92,9 +94,27 @@ export function main() {
+        expect(nestedElBinders[0].nestedProtoView).toBe(nestedElBinders[1].nestedProtoView);
+        done();
+      });
+
```
remove extra newline.

vicb replied:
consistency...

rkirov commented on modules/benchmarks/src/compiler/compiler_benchmark.js:
```diff
@@ -89,14 +96,14 @@ export function main() {
     var cloned = DOM.clone(templateNoBindings);
     cache.clear();
-    compiler.compileAllLoaded(null, annotatedComponent, cloned);
+    compiler.compile(BenchmarkComponent, cloned);
```
I think @tbosch used this to skip the promise path for benchmarks. I am ok with removing it given the extra cost of keeping the fast pass when templateLoader is there too. So just a heads up to Tobias.

vicb replied:
Compiler.compile is now always async, there is no more sync path.

tbosch committed:
refactor(build): explicitly mention `src` folder in imports

Export files are now directly under the module folder,
e.g. `core/core.js`. With this, an import like `core/core`
won’t need a path mapping (e.g. via `System.paths`) any more.
This adds the `src` folder to all other import statements as well.

vsavkin committed:
feat(compiler): add BindingPropagationConfig to the list of pre-built objects

vsavkin committed:
cleanup

mhevery commented:
- [ ] I don't see tests which shows how compiler loads the templates. I would expect to see some kind of mock.
- [ ] We talked about switching to XHR from `<link>`
- [ ]  We need to have a way to mock out the Template loading XHRs.

vsavkin committed:
feat(forms): initial implementation of forms

kevmoo committed:
chore(clean-up): remove unused test method

btford committed:
fix(zones): update to v0.4.0

Closes #487

Commit:
feat(template): add bang syntax shortcut

Closes #522

rkirov commented:
merged (after a monster of a rebase :))

vicb commented:
@rkirov thanks !
